experimental/dispatch_profiler/matmul.py

```diff
@@ -0,0 +1,590 @@
+       1) Generates matmul operation inputs as np.array for a desired 
+          distribution.
+       2) Runs the matmul reference operation in np.matmul.
+       3) Generates the matmul operation expected output as np.array.
+       4) Additional, generate input and output filename strings.
+  """
+
+  def __init__(self, matmul_operation, dist_lhs, dist_rhs):
+    self.matmul_operation = matmul_operation
+    # Problem shape.
+    self.M = matmul_operation.problem_shape[0]
+    self.N = matmul_operation.problem_shape[1]
+    self.K = matmul_operation.problem_shape[2]
+
+    # dtype for the input and result matrices.
+    self.dtype_lhs = DataTypeNumPyTag[matmul_operation.lhs.datatype]
+    self.dtype_rhs = DataTypeNumPyTag[matmul_operation.rhs.datatype]
+    self.dtype_result = DataTypeNumPyTag[matmul_operation.result.datatype]
+
+    # Distribution of the input tensors.
+    self.dist_lhs = dist_lhs
+    self.dist_rhs = dist_rhs
+
+    # filename for the lhs, rhs, inital_result, and result matrices.
+    self.filename_lhs = "m{problem_m}xk{problem_k}_"\
+      "{tensor_description}_{dist}_lhs.npy".format(
+      problem_m=self.M,
+      problem_k=self.K,
+      tensor_description=self.matmul_operation.lhs.name(),
+      dist=DistributionName[self.dist_lhs])
+
+    self.filename_rhs = "k{problem_k}xn{problem_n}_"\
+      "{tensor_description}_{dist}_rhs.npy".format(
+      problem_k=self.K,
+      problem_n=self.N,
+      tensor_description=self.matmul_operation.rhs.name(),
+      dist=DistributionName[self.dist_rhs])
+
+    self.reference_filename_result = "m{problem_m}xn{problem_n}_"\
+      "{tensor_description}_reference_result.npy".format(
+      problem_m=self.M,
+      problem_n=self.N,
+      tensor_description=self.matmul_operation.result.name())
+
+  # Generates input data, runs reference numpy.matmul, and save npy files to the output directory.
+  def run_and_save(self, output_dir="."):
+
+    # Generate the input data as np.array for the matmul operation.
+    lhs_np_array = get_np_array(self.matmul_operation.lhs, (self.M, self.K),
+                                self.dist_lhs)
+    rhs_np_array = get_np_array(self.matmul_operation.rhs, (self.K, self.N),
+                                self.dist_rhs)
+
+    # Run the reference np.matmul and generate result np.array.
+    result = np.matmul(lhs_np_array, rhs_np_array)
+
+    # Save the input data as np.array for the matmul operation.
+    np.save(os.path.join(output_dir, self.filename_lhs),\
+             np.array(lhs_np_array, dtype = self.dtype_lhs))
+    np.save(os.path.join(output_dir, self.filename_rhs),\
+             np.array(rhs_np_array, dtype = self.dtype_rhs))
+
+    # Save the expected result as an np.array.
+    np.save(os.path.join(output_dir, self.reference_filename_result),\
+             np.array(result, dtype = self.dtype_result))
+
+
+###############################################################################
+class MatmulOperationLauncher:
```

Comment:
Comment for the entire `MatmulOperationLauncher` class:

Have you considered IREE's python bindings? We provide a prebuilt version of IREE as well as python code to interact with it that you can use with a simple `pip install`: https://openxla.github.io/iree/building-from-source/python-bindings-and-importers/#building-python-bindings

There is a lot of code here reimplementing what our bindings already do. The bindings also take care of the type conversion between python and IREE. Results are returned as numpy arrays with the proper byte formats (float64 vs float32, etc). 

Ideally, any user should be able to just grab this tool and run with it. Having to build and manage IREE separately is a bigger ask. Using the bindings doesn't remove that option, however, as you can use bindings directly from your `iree-build` dir.

This is all the code I need to compile a flatbuffer with my tuning: 

```python
# Compile according to the input type, else just try compiling.
    try:
        flatbuffer_blob = ireec.compile_str(
            module,
            target_backends=[target_device],
            extra_args=args,
            input_type=input_type,
        )
    except CompilerToolError as err:
        flatbuffer_blob = None
        print(f"Compile tool failed: {err}")
        return None, err
```

Similarly, this is how I run `benchmark_module` (note the results are returned as python objects and errors are handled as well):
```python
    try:
        benchmark_results = benchmark_module(
            vm_module, device=device, benchmark_repetitions=benchmark_repetitions, batch_size=benchmark_dispatch_batch_size)
        return benchmark_results, None
    except BenchmarkToolError as err:
        print(f"Benchmark tool failed: {err}")
        return None, err
```

Reply:
I have made a note on it and will share it through a doc soon. I envision to use this tool for iterating over `iree-compile` as one makes performance improvements. Thus, it would be great if we can run `iree-compile` directly through this tool or cmake instead from being it from python bindings. Yes, python bindings will be useful for iree-benchmark-module and iree-run-module. Let us take it up in a separate PR.

Reply:
Thanks for the explaining this and we will handle this in the future patch to go through the python bindings for `iree-run-module`, `iree-run-benchmark`, and `iree-compile`